docs/rfcs/rfc_firestore_immediate_mode_mutations.md

```diff
@@ -0,0 +1,316 @@
+    writes.push({collection: `orgs/${orgId}/sprints`, doc: nextSprintId })
+  }
+
+  uncompleted.forEach((task) =>
+    writes.concat([{
+      collection: `orgs/${orgId}/sprints`,
+      doc: task.sprintId,
+      data: { orderedTaskIds: arrayRemove(task.id) } }]));
+
+  uncompleted.forEach((task) =>
+    writes.concat([{
+      collection: task.path,
+      doc:task.id,
+      data: { sprint: nextSprintId } }]));
+
+  writes.push({
+    collection: `orgs/${orgId}/sprints`,
+    doc: nextSprintId,
+    data: { orderedTaskIds: uncompleted.map(({id}) => arrayUnion(id))} })
+
+  writes.push({
+    collection: `orgs/${orgId}/team`,
+    doc: team.id,
+    data: { currentSprint: sprintId }
+  });
+
+  return writes;
+}
+
+const writeSprintCalculations = ({completed, uncompleted, sprints}) => {
+  const current = sprint[0];
+  const tasks = completed.concat(uncompleted);
+  const computedOnCompletion = {
+    totalEffortEstimated: tasks.reduce((sum, {effort}) => sum += effort, 0),
+    totalEffortCompleted: completed.reduce((sum, {effort}) => sum += effort, 0)
+  }
+  return {
+    collection: sprint.path,
+    doc:sprint.id,
+    data: { computedOnCompletion } };
+}
+
+const writeSprintComplete = ({uncompleted, sprints, now}) => {
+  const current = sprint[0];
+  return {
+    collection: current.path,
+    doc: current.id,
+    data: {
+      orderedTaskIds: uncompleted.map(({id} => arrayRemove(id))),
+      completed: true,
+      updatedAt:now
+    }
+  };
+}
+
+// Without Redux it's just a simple call to the new mutate function.
+
+const reads = {
+  orgId,
+  teamId,
+  now: new Date(),
+  uncompleted: queryUncompletedTasks,
+  completed: queryCompletedTasks,
+  team: queryTeam,
+  sprints: queryNextSprint,
+};
+
+const = writes: [
+  writeUncompleteToNextSprint,
+  writeSprintCalculations,
+  writeCompleteSprint
+];
+
+firestore.mutate({ reads, writes }, { method: 'transaction' });
+
+// -----
+
+// The Write functions here don't really belong in an action message (but neither do thunks).
+// This is much cleaner via Clojurescript where idempotent functions can just be qouted
+// Other ideas are to put it in a middleware (w/ a string to reference the function)
+dispatch({
+  type: 'COMPLETE_SPRINT',
+  payload: {
+    reads: {
+      orgId,
+      teamId,
+      uncompleted: queryUncompletedTasks,
+      completed: queryCompletedTasks,
+      team: queryTeam,
+      sprints: queryNextSprint,
+    },
+    writes: [
+      writeUncompleteToNextSprint,
+      writeSprintCalculations,
+      writeCompleteSprint
+    ],
+    method: 'transaction'
+  }});
+
+```
+
+# Thoughts on the internals of the mutate function
+
+```ts
+export function mutate(operations, { method = null } = {}) {
+  // if there's a default convertor than add to each firestore call
+  let reads,
+    writes = operations;
+  if (operations.reads) {
+    reads = operations.reads;
+    writes = operations.writes;
+  }
+
+  // first run all optimistic updates synchronously
```

Comment:
Wondering if this logic could fit into a redux middleware? I was actually experimenting with [actions which trigger promises in this middleware](https://github.com/prescottprue/redux-firestore/blob/master/src/middleware.js#L83), so it seems like that could be a good place to do an optimistic update with rollback

Reply:
Interesting thought. What are the positive/negatives of using a middleware for this?

In our project we use redux toolkit’s asyncActionCreator in our project. Honestly for us it’s been a bad pattern. Async is the risky code that can be written.

Here’s the reason. Async code, 99% of the time, is asking for data on disk or over the network, right? That means the code will have zero guarantees about what will come back. It’s not idempotent. It’s not encapsulated state. It’s completely outside the coders control. Promises making the result binary of pass/fail makes it more manageable but more branches and more complexity which adds more failure points. 

The better pattern for us has been sync first, async second. Why? Because the coder has full control to guarantee the outcome. You can show the result to the user via a store update then, after all sync operations complete, then kicking off the async, non-deterministic operations. 

The postive of moving it to a middleware that I see could be that it would allow batching multiple actions better. Hopefully that’s a rare case since the mutation API was made to allow the coder to compose multiple operations as one (like a multi-document transaction). 

On the downside it’s mutating a reducer’s store outside of the reducer. Also state is held in memory as a closure (usually through an instance of a promise) making it harder to debug.